core/src/main/java/org/apache/iceberg/IncrementalDataTableScan.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg;
+
+import com.google.common.base.Preconditions;
+import com.google.common.base.Predicate;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Lists;
+import java.util.Collection;
+import java.util.List;
+import org.apache.iceberg.expressions.Expression;
+import org.apache.iceberg.io.CloseableIterable;
+import org.apache.iceberg.util.SnapshotUtil;
+
+class IncrementalDataTableScan extends DataTableScan {
+  private Long fromSnapshotId;
+  private Long toSnapshotId;
+
+  IncrementalDataTableScan(TableOperations ops, Table table, Schema schema,
+                           Expression rowFilter, boolean caseSensitive, boolean colStats,
+                           Collection<String> selectedColumns, ImmutableMap<String, String> options,
+                           Long fromSnapshotId, Long toSnapshotId) {
+    super(ops, table, null, schema, rowFilter, caseSensitive, colStats, selectedColumns, options);
+    Preconditions.checkArgument(fromSnapshotId != toSnapshotId, "fromSnapshotId and toSnapshotId cannot be the same");
+    this.fromSnapshotId = fromSnapshotId;
+    this.toSnapshotId = toSnapshotId;
+  }
+
+  @Override
+  public TableScan asOfTime(long timestampMillis) {
+    throw new UnsupportedOperationException("asOfTime() is not supported in IncrementalScan");
+  }
+
+  @Override
+  public TableScan useSnapshot(long scanSnapshotId) {
+    throw new UnsupportedOperationException("useSnapshot() is not supported in incremental scan");
+  }
+
+  @Override
+  public TableScan appendsBetween(Long newFromSnapshotId, Long newToSnapshotId) {
+    Preconditions.checkArgument(
+        table().snapshot(newFromSnapshotId) != null, "fromSnapshotId: %s does not exist", newFromSnapshotId);
+    Preconditions.checkArgument(
+        table().snapshot(newToSnapshotId) != null, "toSnapshotId: %s does not exist", newToSnapshotId);
+    return new IncrementalDataTableScan(
```

Comment:
Should this ensure that `newFromSnapshot` is an ancestor of `newToSnapshot`?

This probably doesn't need to be done in this commit, but it would be a good follow-up to ensure that the range exists.

Since this is a refinement, it may also be a good idea to make this a subset of the existing selected range. That is, both `newFromSnapshotId` and `newToSnapshotId` must be in the existing range of `fromSnapshotId` to `toSnapshotId`.

Putting it another way, when I create a scan using `appendsBetween(A, C).appendsBetween(B, D)`, what should the behavior be? I'd say that is concerning because D is outside the original range. Probably a good idea to fail instead.

Reply:
Thankyou, I'll file ticket for followup!